module/VuFind/src/VuFind/UrlShortener/Database.php

```diff
@@ -133,25 +92,33 @@ protected function base62Decode($base62Number)
     public function shorten($url)
     {
         $path = str_replace($this->baseUrl, '', $url);
-        $this->table->insert(['path' => $path]);
-        $id = $this->table->getLastInsertValue();
+        $hash = hash(static::HASH_ALGO, $path . $this->salt);
+        $shorthash = substr($hash, 0, 9);
+        $results = $this->table->select(['hash' => $shorthash]);
+
+        // this should almost never happen - we then return the existing hash
+        if ($results->count() == 0) {
```

Comment:
I realize that a hash collision is pretty unlikely, but since we're only using a substring of the hash, it seems like it might be possible for two different paths to lead to the same shorthash. In that case, we don't want to return the wrong path in response to a hash! Would it be better to add an additional check here to see if the path matches, and if it doesn't, perhaps we could repeat the process with a longer hash segment (10 characters, then 11, then 12, etc., until we find something unique).

Reply:
I figured this out, but dropped it because in our usage scenario there will never be so many links that collisions are likely. The email feature is not used very intense. But You are right If we want to do it perfect, then we need to save the full hash and increase the number of characters, first.

Reply:
I've added some smarts to hopefully avoid this possibility.